Change process spawning to inherit the parent's signal mask by default

sunshowers opened the pull request:
Previously, the signal mask was always reset when a child process is
started. This breaks tools like `nohup` which expect `SIGHUP` to be
blocked for all transitive processes.

With this change, the default behavior changes to inherit the signal mask.

This also changes the signal disposition for `SIGPIPE` to only be changed if the `#[unix_sigpipe]` attribute isn't set.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

sunshowers commented:
r? @thomcc

sunshowers commented:
(I don't know of a way to represent the stack cleanly in GitHub, unlike in Differential or other tools.)

thomcc commented:
Ideally it would be cleanly separated from the other (I know this is a bit of a pain). That would make it easier to FCP and land on its own.

sunshowers commented:
It's not really possible to cleanly separate. Best one could do is to change the ordering so that #100737 depends on this rather than the other way round.

To avoid duplicated work I would be happy to do that after the FCP regarding this is approved.

thomcc commented:
Sure. I'll assign to someone who can start an FCP for this (I really need to get to https://github.com/rust-lang/rfcbot-rs/issues/306... maybe on Monday...). For their context, this is based on this comment from @zackw: https://github.com/rust-lang/rust/pull/100737#issuecomment-1222672934

r? @joshtriplett

sunshowers commented:
Is there anything blocking this? @joshtriplett @thomcc

Once this is accepted I'd be happy to change the ordering of the PRs and resolve merge conflicts, etc.

joshtriplett commented on library/std/src/os/unix/process.rs:
```diff
@@ -179,6 +179,103 @@ pub trait CommandExt: Sealed {
+    /// Command::new("sleep")
+    ///     .arg("10")
+    ///     // On most platforms, SIGINT is signal 2.
+    ///     .block_signal(2)?
```
The comment should say something like "prefer libc::SIGINT if available"

sunshowers replied:
(This is no longer relevant as of this change -- I'll fix #100737 up with it.)

joshtriplett commented:
I think the handling for SIGPIPE needs some updates, for two reasons:

- This should take https://github.com/rust-lang/rust/pull/97802 into account, rather than unconditionally unblocking.
- If the user uses block_signal on SIGPIPE, we should respect that and not override it.

joshtriplett commented:
@rustbot author

Enselic commented:
> This should take [Support `#[unix_sigpipe = "inherit|sig_dfl"]` on `fn main()` to prevent ignoring `SIGPIPE` #97802](https://github.com/rust-lang/rust/pull/97802) into account, rather than unconditionally unblocking.

Would you mind elaborating a bit on that please? In my mind that attribute only affects what the runtime does to SIGPIPE before fn main() is invoked. I don't think it should affect the spawning of child processes? (Except indirectly)

joshtriplett commented:
"except indirectly" is exactly my concern here. Right now, and with this patch, Rust resets SIGPIPE to default in the child process. I think if the user sets the sigpipe disposition with that attribute, we should allow that to be inherited unless overridden.

Enselic commented:
Hmm, in that case we should probably make `#[unix_sigpipe = "..."]` a module attribute rather than an attribute on `fn main()`, because otherwise the behaviour of inheritance of parent signal masks can't be configured for library code. But if we make it an attribute that can be used with libraries, how do we handle when two different libraries that are in use choose different values on the attribute?

To me, these problems suggest that the `#[unix_sigpipe = "..."` "belongs" to `fn main()`, since without an `fn main()`, no code is generated that can configure `SIGPIPE` before `fn main()` is invoked.

I think it is useful to think about what we would do if the Rust runtime would never have started to touch `SIGPIPE` in 2014. If that would have been the case, how would we have written the code in this PR? And I think the answer to that is "we do not treat `SIGPIPE` any different than any other signal", so that seems to me to be what we should do also today here.

But I am of course open to be convinced otherwise.

joshtriplett commented:
@Enselic I agree, and I wish we could do that, but I think that would break backwards compatibility at the moment.

I'm proposing that if something sets the attribute *at all*, we treat that as an opt in to not touching the signal *here*.

sunshowers commented:
@joshtriplett Some questions:

1.

How should `reset_sigpipe`: https://github.com/rust-lang/rust/blob/8c6ce6b91b172f77c795a74bfeaf74b865146b3f/library/std/src/sys/unix/mod.rs#L156-L181

    communicate with process spawning code? My first thought was to use a static `OnceLock`, but I'm wondering if it'll run into issues given how early in the process this happens. Do you have other/better suggestions?

2. Note that this PR touches `setsigmask` (AIUI a mask of signals currently blocked from delivery, though can be delivered in the future), not `setsigdefault` (signals that are reset to default signal handling, but which can still potentially be temporarily blocked by my understanding). If `block_signal` is called, that doesn't directly affect `setsigdefault`, only `setsigmask`.
  a. How should the behavior of `setsigdefault` change under `block_signal`?
  b. If `SIGPIPE` is currently *blocked* in this process, and `unblock_signal` is called, what should happen to the behavior of `setsigdefault`?

3. What's the best way to test this behavior? It seems like there are a number of different combinations involved here.

sunshowers committed:
Change process spawning to inherit the parent's signal mask by default

Previously, the signal mask is always reset when a child process is
started. This breaks tools like `nohup` which expect `SIGHUP` to be
blocked.

With this change, the default behavior changes to inherit the signal mask.

This also changes the signal disposition for `SIGPIPE` to only be
changed if the `#[unix_sigpipe]` attribute isn't set.

sunshowers commented:
Changes since the last version:

1. Made it so that this is now free-standing and no longer dependent on #100737.
2. Per @joshtriplett's command, changed the code so that `SIGPIPE` is no longer set if `#[unix_sigpipe]` is specified.
3. Added documentation to the unix-sigpipe unstable feature about 2, noting in particular that there's now a difference between not specifying it and specifying `"sig_ign"`.

The `block_signal` considerations can be deferred to #100737 -- they're no longer relevant with this version.

I still don't know how to test the `SIGPIPE` change yet, unfortunately.

sunshowers commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

sunshowers commented on library/std/src/sys/unix/process/process_unix.rs:
```diff
@@ -326,30 +326,26 @@ impl Command {
+            //
+            // #[unix_sigpipe] is an opportunity to change the default here.
+            if !unix_sigpipe_attr_specified() {
+                #[cfg(target_os = "android")] // see issue #88585
+                {
+                    let mut action: libc::sigaction = mem::zeroed();
+                    action.sa_sigaction = libc::SIG_DFL;
+                    cvt(libc::sigaction(libc::SIGPIPE, &action, ptr::null_mut()))?;
+                }
```
Is the behavior on Android correct? I'm not sure.

sunshowers commented:
OK so @joshtriplett @Enselic I've spent a couple of hours on this and I still don't know how to write an integration test for this using compiletest.

The thing I would like to test is: ensure that the process spawning code doesn't touch the `SIGPIPE` signal handler if `#[unix_sigpipe]` is specified. The closest example I've found is https://github.com/rust-lang/rust/blob/4a14677239bbd9a3892d165ee24513c1a05820e4/src/test/ui/process/process-sigpipe.rs#L22-L36. So I tried writing a variant of this test that sets `#[unix_sigpipe = "sig_ign"]` etc.

How do I actually test the behavior? Well, some options I can think of:
1. **Use a shell script.** How? The `yes | head` has a pretty clear error case if we want to test `SIG_DFL`. However, if we want to test `SIG_IGN`, this just produces `yes: standard output: Broken pipe` on my machine, which is text I don't think we can rely on to be stable.
2. **Use a Rust binary.** This would be ideal. But the only examples I can find execute the current binary, which is clearly not suitable since it'll be influenced by the `#[unix_sigpipe]` directive. We'd want to have a Rust test execute *another* Rust binary, and I don't know if compiletest supports that.
3. **Use Python?** I don't know if this is a feasible option.

As a meta comment, I'm still not sure what this has to do with my change, which is to inherit the signal *mask*, and is unrelated to signal *handlers*. I think this should be a separate PR that is unrelated to this.

joshtriplett commented:
The current approach looks good to me, and I think it'd be reasonable to work on a test separate from this PR.

Enselic commented on compiler/rustc_session/src/config/sigpipe.rs:
```diff
@@ -16,7 +16,10 @@ pub const SIG_IGN: u8 = 2;
 /// <https://github.com/rust-lang/rust/issues/62569>.
 #[allow(dead_code)]
-pub const DEFAULT: u8 = SIG_IGN;
+pub const DEFAULT: u8 = 0;
```
I agree `0` is a good value for `DEFAULT`, but can we perhaps move it to the top of the file so that the constants becomes sequential? Or maybe change the value to `4`, but that seems less attractive to me.

We should also update

https://github.com/rust-lang/rust/blob/bb5a01617589b5e3ece5a36435fc285bfd13c9a4/library/std/src/rt.rs#L92

to not say "3 values", but instead say "4 values". Or perhaps even better: rephrase it to not specify an exact count.

sunshowers replied:
Done, thanks.

Enselic commented on library/std/src/sys/unix/mod.rs:
```diff
@@ -166,21 +167,45 @@ pub unsafe fn init(argc: isize, argv: *const *const u8, sigpipe: u8) {
                 pub const INHERIT: u8 = 1;
                 pub const SIG_IGN: u8 = 2;
                 pub const SIG_DFL: u8 = 3;
+                pub const DEFAULT: u8 = 0;
```
(Also move this to the top if you move the definition to the top)

Enselic commented on library/std/src/sys/unix/mod.rs:
```diff
@@ -166,21 +167,45 @@ pub unsafe fn init(argc: isize, argv: *const *const u8, sigpipe: u8) {
+                sigpipe::INHERIT => (true, None),
+                sigpipe::SIG_IGN => (true, Some(libc::SIG_IGN)),
+                sigpipe::SIG_DFL => (true, Some(libc::SIG_DFL)),
+                sigpipe::DEFAULT => (false, Some(libc::SIG_IGN)),
```
(Also move this to the top if you move the definition to the top)

sunshowers replied:
Done.

Enselic commented on src/doc/unstable-book/src/language-features/unix-sigpipe.md:
```diff
@@ -36,7 +36,7 @@ hello world
 Set the `SIGPIPE` handler to `SIG_IGN` before invoking `fn main()`. This will result in `ErrorKind::BrokenPipe` errors if you program tries to write to a closed pipe. This is normally what you want if you for example write socket servers, socket clients, or pipe peers.
 
-This is what libstd has done by default since 2014. Omitting `#[unix_sigpipe = "..."]` is the same as having `#[unix_sigpipe = "sig_ign"]`.
+This is what libstd has done by default since 2014. (However, see the note on child processes below). Omitting `#[unix_sigpipe = "..."]` is the same as having `#[unix_sigpipe = "sig_ign"]`.
```
We should remove the phrase

> Omitting `#[unix_sigpipe = "..."]` is the same as having `#[unix_sigpipe = "sig_ign"]`.

now, right?

sunshowers replied:
Yep, done!

sunshowers commented:
OK, review comments addressed. I think this is ready for FCP and to land :)

bors commented:
:umbrella: The latest upstream changes (presumably #102051) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3138281961/jobs/5097435494) [(plain)](https://github.com/rust-lang/rust/commit/788d56dc83f8a1938c29bc1b9f1de11b5038ce8d/checks/8580494753/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
stdout: none
--- stderr -------------------------------
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`', /checkout/src/test/ui/attributes/unix_sigpipe/auxiliary/sigpipe-utils.rs:26:9
------------------------------------------



```

</details>

joshtriplett commented:
This looks good to me. Since this changes user-visible behavior, and *can't* be put behind a nightly feature gate, I think we need to FCP this.

@rfcbot merge

Xiretza commented:
Failed in rollup: https://github.com/rust-lang/rust/pull/103247#issuecomment-1284310551

sunshowers commented:
Thanks. I've addressed the failures -- I think this is ready to land again.

thomcc commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit 96baf3231c2716f0b97d4ef7b8962c6090a64bf5 has been approved by `thomcc`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=thomcc 96baf3231c2716f0b97d4ef7b8962c6090a64bf5 -->
<!-- homu: {"type":"Approved","sha":"96baf3231c2716f0b97d4ef7b8962c6090a64bf5","approver":"thomcc","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:v: @sunshowers can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"thomcc","delegate":"sunshowers"} -->

bors commented:
:hourglass: Testing commit 96baf3231c2716f0b97d4ef7b8962c6090a64bf5 with merge 910d63f3e61bb4fcb89dd458cc1132b8fb1d9783...
<!-- homu: {"type":"BuildStarted","head_sha":"96baf3231c2716f0b97d4ef7b8962c6090a64bf5","merge_sha":"910d63f3e61bb4fcb89dd458cc1132b8fb1d9783"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3285857599/jobs/5413471286)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3285857599/jobs/5413471286","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`dist-android`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3285857599/jobs/5413347088) [(plain)](https://github.com/rust-lang-ci/rust/commit/910d63f3e61bb4fcb89dd458cc1132b8fb1d9783/checks/8996384143/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling addr2line v0.16.0
[RUSTC-TIMING] addr2line test:false 0.457
[RUSTC-TIMING] gimli test:false 4.263
[RUSTC-TIMING] object test:false 4.488
error: function `sigemptyset` is never used
   |
   |
42 |         pub unsafe fn sigemptyset(set: *mut libc::sigset_t) -> libc::c_int {
   |
   |
   = note: `-D dead-code` implied by `-D warnings`
[RUSTC-TIMING] std test:false 3.681
error: could not compile `std` due to previous error
Build completed unsuccessfully in 0:10:47
```

</details>

sunshowers commented:
@bors r+

bors commented:
:pushpin: Commit b046dabb80020fc595b84a0702c2e1831f132d06 has been approved by `sunshowers`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=sunshowers b046dabb80020fc595b84a0702c2e1831f132d06 -->
<!-- homu: {"type":"Approved","sha":"b046dabb80020fc595b84a0702c2e1831f132d06","approver":"sunshowers","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit b046dabb80020fc595b84a0702c2e1831f132d06 with merge 28d0c8d884c62623d875a9f2e27d2da147a376ff...
<!-- homu: {"type":"BuildStarted","head_sha":"b046dabb80020fc595b84a0702c2e1831f132d06","merge_sha":"28d0c8d884c62623d875a9f2e27d2da147a376ff"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3292746353/jobs/5428915367)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3292746353/jobs/5428915367","builder_name":"checks-actions"} -->

sunshowers commented:
@bors r+

bors commented:
:pushpin: Commit a52c79e859142c1cd5c0c5bdb73f16b754e1b98f has been approved by `sunshowers`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=sunshowers a52c79e859142c1cd5c0c5bdb73f16b754e1b98f -->
<!-- homu: {"type":"Approved","sha":"a52c79e859142c1cd5c0c5bdb73f16b754e1b98f","approver":"sunshowers","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-log-analyzer commented:
The job **`dist-various-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3292746353/jobs/5428475391) [(plain)](https://github.com/rust-lang-ci/rust/commit/28d0c8d884c62623d875a9f2e27d2da147a376ff/checks/9016751633/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  |
5 | use crate::ptr;
  |     ^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
[RUSTC-TIMING] std test:false 3.335
warning: `std` (lib) generated 1 warning
error: could not compile `std` due to previous error; 1 warning emitted
Build completed unsuccessfully in 0:16:55
```

</details>

thomcc commented:
FYI if you want to run the CI checks without waiting for bors to get around to trying it, see https://rustc-dev-guide.rust-lang.org/tests/ci.html?highlight=ci#using-ci-to-test for a guide. (Just be sure not to land any CI changes, of course)

sunshowers commented:
Thanks. Apparently this last comment was just from my previous land attempt. Hopefully this one goes smoother.

bors commented:
:hourglass: Testing commit a52c79e859142c1cd5c0c5bdb73f16b754e1b98f with merge 57e2c06a8df3187980801962015a88657abd033d...
<!-- homu: {"type":"BuildStarted","head_sha":"a52c79e859142c1cd5c0c5bdb73f16b754e1b98f","merge_sha":"57e2c06a8df3187980801962015a88657abd033d"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3299523298/jobs/5444880270)
Approved by: sunshowers
Pushing 57e2c06a8df3187980801962015a88657abd033d to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"sunshowers","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3299523298/jobs/5444880270"},"merge_sha":"57e2c06a8df3187980801962015a88657abd033d"} -->

rust-timer commented:
Finished benchmarking commit (57e2c06a8df3187980801962015a88657abd033d): [comparison URL](https://perf.rust-lang.org/compare.html?start=ba9d01be67a97dd6762935762172e37ee22083ae&end=57e2c06a8df3187980801962015a88657abd033d&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ba9d01be67a97dd6762935762172e37ee22083ae&end=57e2c06a8df3187980801962015a88657abd033d&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ba9d01be67a97dd6762935762172e37ee22083ae&end=57e2c06a8df3187980801962015a88657abd033d&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=ba9d01be67a97dd6762935762172e37ee22083ae&end=57e2c06a8df3187980801962015a88657abd033d&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

passcod commented:
Just some feedback that this [broke cargo-watch](https://github.com/watchexec/cargo-watch/issues/249) as we set sigmask on the parent process and relied on the sigmask reset for children to run them as the user expected.

sunshowers commented:
You can fix it locally by temporarily blocking the signals in the spawning thread. See https://github.com/nextest-rs/nextest/blob/2505a731eadd7b7bc7355a53e753a05d822a358d/nextest-runner/src/double_spawn.rs#L167 for an example.

passcod commented:
It just seems weird because the previous thing was deliberately done: https://github.com/rust-lang/rust/pull/25784
